### ISSUE TITLE
New ICE from modified example code: Uninferred types/regions in `_`

### ISSUE THREAD

**Issue by kyleheadley** [rust-lang/rust#43478]
I tried to instantiate the types in #36839 with a `Vec` and received the original ICE.

[playground](https://play.rust-lang.org/?gist=4da142b75be06065ebaf2af68c0f575b&version=nightly)

```
pub trait Foo {
    type Bar;
}

pub trait Broken {
    type Assoc;
    fn broken(&self)->Vec<<Self::Assoc as Foo>::Bar> where Self::Assoc: Foo;
}

impl<T> Broken for T {
    type Assoc = ();
    fn broken(&self)->Vec<<Self::Assoc as Foo>::Bar> where Self::Assoc: Foo {
        Vec::new()
    }
}

fn main() { 
    let _m: &Broken<Assoc=()> = &();
}
```

What's the expected behavior here? A compiler error on trait impl or a compiler error if the fn is called?

**Comment by estebank** [comment]
No longer reproducible.
